amazon-connect/amazon-connect-chatjs
# [Description/Design] :add advancedLog to the websocketManager

-add advancedLog config
-fix initialization of websocket logger to log messages and match ChatJS logger
-add ability to call GetTranscript after web socket connection has ended. 
-add ability to re-connect to web socket after connection has ended

*Description of changes:*

The logs exposed from WebSocketManager introduce too many logs and will quickly fill up the log bucket so every consumer either needs to manually filter specific logs or subset of logs. To resolve this we created a new sub-set using advancedLogs which can be selected/enabled by Consumers of ChatJS & WebsocketManager to gather minimal logs from production which are required for debugging.

*How to Consume new advancedLog*
https://github.com/amazon-connect/amazon-connect-chat-interface/blob/master/src/utils/log.js

```
  const customizedLogger = {
    debug: (data) => {// customize logger function here}, 
    info: (data) => {// customize logger function here}, 
    warn(data) {} =>  {// customize logger function here}, 
    error: (data) => {// customize logger function here}, 
  }

export const config = {
  loggerConfig: {
    // logger/customizedLogger: Add your logger here
    customizedLogger: customizedLogger,
    level: connect.LogLevel.DEBUG,
    useDefaultLogger: true,
    advancedLogWriter: 'error',
    //useDefaultLogger uses window.console
    //advancedLogWriter: can have `info`,`warn`,`debug`,`error` - default value is `warn`
  },
};
```

connect.ChatSession.setGlobalConfig(config);

Note: advancedLogWrite exposes minimum set of logs needed for websocket production debugging. So, its recommended to save logs exposed by advancedLogWriter while debugging a production connectivity issue.

*example logs:*
```
amazon-connect-chat-interface.js:2 [2022-08-10T17:19:42.465Z][ADVANCED_LOG]AMZ_WEB_SOCKET_MANAGER: Reset Websocket state 
e.warn @ amazon-connect-chat-interface.js:2
amazon-connect-chat-interface.js:2 [2022-08-10T17:19:42.465Z][ADVANCED_LOG]AMZ_WEB_SOCKET_MANAGER: Fetching new WebSocket connection configuration 
```
-----------

*Testing:*

- updated WebsocketManager to call terminateWebSocketManager when network is offline
- build the changes into ChatJS
- add onConnectionBroken callback listener in ChatInterface
```
this.client.onConnectionBroken(() => {
    if (!transcriptPollingInterval) {
      transcriptPollingInterval = setInterval(() => {
        // Retrieve transcript and trigger rendering new transcript items to UI
        // ChatInterface ref: https://github.com/amazon-connect/amazon-connect-chat-interface/blob/master/src/components/Chat/ChatSession.js#L360
        this.loadPreviousTranscript(); 
        // should there be a network online check here?
      }, 10000);
    }
  });
```
- copy the dist/ package to public-chat-examples application and launch Custom-Chat-Widget .

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
